File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-1.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        
        l = 1
        r = len(nums)

        res = nums[0]

        while l <= r:

            m = (l + r)//2 % len(nums)

            res = min(res, nums[m])

            if nums[m] > nums[len(nums)-1]:
                l = m + 1
            else:
                r = m - 1
        return res

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> res(nums.size());
        vector<int> pre(nums.size());
        vector<int> post(nums.size());

        pre[0] = 1;
        post[nums.size()-1] = 1;
        for(int i = 1; i < nums.size(); i++)
        {
            pre[i] = nums[i-1] * pre[i-1];
        }
        for(int i = nums.size()-2; i >= 0; i--)
        {
            post[i] = nums[i+1] * post[i+1];
        }
        for(int i = 0; i < nums.size(); i++)
        {
            res[i] = pre[i] * post[i];
        }
        return res;
    }
};
